.changeset/long-icons-pull.md

```diff
@@ -0,0 +1,37 @@
+---
+'@xstate/react': patch
+---
+
+Correct reference to the machine in `useInterpret` hook.
+Make sure a warning is shown in the console in the development mode when machine reference is updated during the hook lifecycle. This usually happens when machine options are dependent on external values and they're passed via `withConfig`.
+
+```js
+const machine = createMachine({
+  initial: 'foo',
+  context: { id: 1 },
+  states: {
+    foo: {
+      on: {
+        CHECK: {
+          target: 'bar',
+          cond: 'hasOverflown'
+        }
+      }
+    },
+    bar: {}
+  }
+});
+
+const [id, setId] = useState(1);
+const [current, send] = useMachine(
+  machine.withConfig({
+    guards: {
+      hasOverflown: () => id > 1 // id is a reference to an outside value
+    }
+  })
+);
+
+// later when id updates
+setId(2);
+// Now the reference to `machine.withConfig` from above is updated but in the guard, the id value is still stale (id=1).
+```
```

Comment:
It would be great if we could also mention how one can fix this here, add a note to the docs about this and maybe even add a link to that note in the warning itself

Reply:
Good idea. I'm updating the docs in a separate PR. in that same PR, I'll also refer to it here. sounds good?

Reply:
Slightly unsure about the exact shape of that other PR - but if you prefer to make those changes in a separate PR, then it's fine for me.

Reply:
it'll be a doc-only change. Let's take it there.